pkg/solana/chainreader/bindings.go

```diff
@@ -13,167 +14,420 @@ import (
-		for _, rb := range nbs {
-			rb.SetModifier(modifier)
+func (r *bindingsRegistry) initAddressSharing(addressShareGroups [][]string) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.addressShareGroups = make(map[string]*addressShareGroup)
+
+	for _, group := range addressShareGroups {
+		shareGroup := &addressShareGroup{
+			address: solana.PublicKey{},
+			group:   group,
+		}
+
+		for _, namespace := range group {
+			if _, alreadySharesAddress := r.addressShareGroups[namespace]; alreadySharesAddress {
+				return fmt.Errorf("namespace %q can't share address with two different groups", namespace)
+			}
+
+			r.addressShareGroups[namespace] = shareGroup
 		}
 	}
+
+	return nil
+}
+
+func (r *bindingsRegistry) getShareGroup(nameSpace string) (*addressShareGroup, bool) {
+	shareGroup, sharesAddress := r.addressShareGroups[nameSpace]
+	if !sharesAddress {
+		return nil, false
+	}
+
+	return shareGroup, sharesAddress
 }
 
-func (b *bindingsRegistry) handleAddressSharing(boundContract *types.BoundContract) error {
-	shareGroup, isInAGroup := b.getShareGroup(boundContract.Name)
+func (r *bindingsRegistry) handleAddressSharing(boundContract *types.BoundContract) error {
+	shareGroup, isInAGroup := r.getShareGroup(boundContract.Name)
 	if !isInAGroup {
 		return nil
 	}
 
-	shareGroup.mux.Lock()
-	defer shareGroup.mux.Unlock()
-
 	// set shared address to the binding address
-	if shareGroup.address.IsZero() {
+	if shareGroup.getAddress().IsZero() {
 		key, err := solana.PublicKeyFromBase58(boundContract.Address)
 		if err != nil {
 			return err
 		}
-		b.addressShareGroups[boundContract.Name].address, shareGroup.address = key, key
-	} else if boundContract.Address != shareGroup.address.String() && boundContract.Address != "" {
-		return fmt.Errorf("namespace: %q shares address: %q with namespaceBindings: %v and cannot be bound with a new address: %s", boundContract.Name, shareGroup.address, shareGroup.group, boundContract.Address)
+
+		shareGroup.setAddress(key)
+	} else if boundContract.Address != shareGroup.getAddress().String() && boundContract.Address != "" {
+		return fmt.Errorf("namespace: %q shares address: %q with namespaceBindings: %v and cannot be bound with a new address: %s", boundContract.Name, shareGroup.getAddress(), shareGroup.group, boundContract.Address)
+	}
+
+	// side-effect of updating the bound contract address to group-shared address
+	boundContract.Address = shareGroup.getAddress().String()
+
+	return nil
+}
+
+type namespaceBinding struct {
+	// static data
+	name string
+
+	// dynamic thread-safe data
+	mu      sync.RWMutex
+	readers map[string]readBinding
+	bound   map[solana.PublicKey]bool
+}
+
+func newNamespaceBinding(namespace string) *namespaceBinding {
+	return &namespaceBinding{
+		name:    namespace,
+		readers: make(map[string]readBinding),
+		bound:   make(map[solana.PublicKey]bool),
 	}
+}
+
+func (b *namespaceBinding) SetCodecs(codec types.RemoteCodec) {
+	b.mu.RLock()
```

Comment:
Shouldn't we use a write lock for the set method?

Reply:
The `*namespaceBinding` is only doing a read on `b.readers`.

Reply:
Never mind, realized the set is on the reader so you're right only needing a read lock here and below